presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -386,6 +399,42 @@ public PrestoSparkTaskExecutorFactory(
                 allocationTrackingEnabled,
                 false);
 
+        final double memoryRevokingThreshold = getMemoryRevokingThreshold(session);
+        if (isSpillEnabled(session)) {
+            memoryPool.addListener((pool, queryId, totalMemoryReservationBytes) -> {
+                if (totalMemoryReservationBytes > queryContext.getPeakNodeTotalMemory()) {
+                    queryContext.setPeakNodeTotalMemory(totalMemoryReservationBytes);
+                }
+                if (totalMemoryReservationBytes > maxTotalMemory.toBytes()) {
+                    throw exceededLocalTotalMemoryLimit(
+                            maxTotalMemory,
+                            queryContext.getAdditionalFailureInfo(totalMemoryReservationBytes, 0) +
+                                    format("Total reserved memory: %s, Total revocable memory: %s",
+                                            succinctBytes(pool.getQueryMemoryReservation(queryId)),
+                                            succinctBytes(pool.getQueryRevocableMemoryReservation(queryId))));
+                }
+                if (totalMemoryReservationBytes > pool.getMaxBytes() * memoryRevokingThreshold && memoryRevokePending.compareAndSet(false, true)) {
+                    memoryUpdateExecutor.execute(() -> {
```

Comment:
I initially tried synchronously, local test hangs.  `MemoryRevokingSchedule` also trigger spilling asynchronously.

Reply:
When an operator starts to spill,  can we stop execute this listener?   I tried to use a flag to only call `operatorContext.requestMemoryRevoking();` once, however, it was sent too early and the drivers in `PipelineContext` are not populated yet, thus spilling is not set successfully.    What is the overhead of keep running this listener when memory is above threshold?  Not that much, if an operator is set as spill, `operatorContext.requestMemoryRevoking();`  will do nothing and return.